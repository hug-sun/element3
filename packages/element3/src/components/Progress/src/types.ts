export interface LevelColor {
  color: string
  percentage: number
}

export type FnColor = (percentage: number) => string

// export type AnyColor = string | string[] | LevelColor[] | FnColor
export type AnyColor = string | FnColor | unknown[]

export type FnFormat = (percentage: number) => string

export interface StrokeStyle {
  strokeDasharray: string
  strokeDashoffset: string
  transition?: string
}

export interface ProgressProps {
  percentage?: number
  color?: AnyColor
  status?: string
  strokeWidth?: number
  type?: string
  format?: FnFormat
  showText?: boolean
  textInside?: boolean
  width?: number
}
