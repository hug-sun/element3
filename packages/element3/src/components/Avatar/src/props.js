import { isString, isNumber } from '../../../utils/types'

export const props = {
  icon: {
    type: String
  },
  size: {
    type: [Number, String],
    default: 'large',
    validator(val) {
      if (isString(val)) {
        return ['large', 'medium', 'small'].includes(val)
      }
      return isNumber(val)
    }
  },
  shape: {
    type: String,
    default: 'circle',
    validator(val) {
      return ['circle', 'square'].includes(val)
    }
  },
  src: {
    type: String
  },
  alt: {
    type: String
  },
  srcSet: {
    type: String
  },
  fit: {
    type: String,
    default: 'cover'
  },
  error: {
    type: Function
  }
}
