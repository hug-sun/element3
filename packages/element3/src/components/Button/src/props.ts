import { ButtonType, ButtonSize, ButtonNativeType } from './types'
import { PropType } from 'vue'

export const props = {
  size: {
    type: String as PropType<ButtonSize>,
    validator(val: string): boolean {
      return ['medium', 'small', 'mini', ''].includes(val)
    }
  },
  type: {
    type: String as PropType<ButtonType>,
    validator(val: string): boolean {
      return [
        'primary',
        'success',
        'warning',
        'danger',
        'info',
        'text'
      ].includes(val)
    }
  },
  nativeType: {
    type: String as PropType<ButtonNativeType>,
    default: 'button'
  },
  plain: Boolean,
  round: Boolean,
  circle: Boolean,
  loading: Boolean,
  disabled: Boolean,
  icon: String
}
