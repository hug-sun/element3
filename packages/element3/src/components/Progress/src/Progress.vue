<template>
  <div :class="rootClass">
    <div class="el-progress-bar" v-if="type === 'line'">
      <div class="el-progress-bar__outer" :style="barOuterStyle">
        <div class="el-progress-bar__inner" :style="barStyle">
          <div class="el-progress-bar__innerText" v-if="showText && textInside">
            {{ content }}
          </div>
        </div>
      </div>
    </div>
    <div class="el-progress-circle" v-else :style="circleStyle">
      <svg :viewBox="viewBox">
        <path
          class="el-progress-circle__track"
          stroke="#e5e9f2"
          :stroke-width="svgStrokeWidth"
          fill="none"
          :d="svgPathD"
          :style="trailPathStyle"
        ></path>
        <path
          class="el-progress-circle__path"
          :stroke="svgStrokeColor"
          :stroke-width="svgStrokeWidth"
          :stroke-linecap="strokeLinecap"
          fill="none"
          :d="svgPathD"
          :style="arcPathStyle"
        ></path>
      </svg>
    </div>
    <div
      class="el-progress__text"
      v-if="showText && !textInside"
      :style="textStyle"
    >
      <template v-if="!status">{{ content }}</template>
      <i v-else :class="iconClass"></i>
    </div>
  </div>
</template>

<script lang="ts">
import { computed, ComputedRef, defineComponent, Ref, toRefs, unref } from 'vue'
import {
  props,
  statusValid,
  autoFixPercentage,
  STATUS_SETTING,
  SVG_VIEW_BOX,
  calcRelativeSvgSize,
  generateSvgPathD,
  genTrailPathStyle,
  calcPerimeter,
  calcSvgRadius,
  genArcPathStyle,
  getSvgStrokeColor,
  getColorBy
} from '../src/props'
import type { AnyColor, FnFormat } from './types'

export default defineComponent({
  name: 'ElProgress',
  props,
  setup(props) {
    const {
      percentage,
      format,
      color,
      strokeWidth,
      type,
      status,
      showText,
      textInside,
      width
    } = toRefs(props)
    const barStyle = useBarStyle(percentage, color)
    const barOuterStyle = useBarOuterStyle(strokeWidth)
    const content = useContent(format, percentage)
    const iconClass = useIconClass(status, type)
    const rootClass = useRootClass(type, status, showText, textInside)
    const circleStyle = useCircleStyle(width)
    const viewBox = SVG_VIEW_BOX
    const svgStrokeWidth = useSvgStrokeWidth(strokeWidth, width)
    const svgPathD = useSvgPathD(svgStrokeWidth, type)
    const trailPathStyle = useTrailPathStyle(svgStrokeWidth, type)
    const arcPathStyle = useArcPathStyle(svgStrokeWidth, percentage, type)
    const svgStrokeColor = useSvgStrokeColor(status, color, percentage)
    const textStyle = useTextStyle(type, width)
    return {
      barStyle,
      barOuterStyle,
      content,
      iconClass,
      rootClass,
      circleStyle,
      viewBox,
      svgPathD,
      svgStrokeWidth,
      trailPathStyle,
      arcPathStyle,
      svgStrokeColor,
      textStyle
    }
  }
})

const useRootClass = (
  type: Ref<string>,
  status: Ref<string>,
  showText: Ref<boolean>,
  textInside: Ref<boolean>
) => {
  return computed(() => {
    const valType = unref(type)
    const valStatus = unref(status)
    const valShowText = unref(showText)
    const valTextInside = unref(textInside)
    const statusClass =
      valStatus && statusValid(valStatus) ? `is-${valStatus}` : ''
    return [
      'el-progress',
      `el-progress--${valType}`,
      statusClass,
      {
        'el-progress--without-text': !valShowText,
        'el-progress--text-inside': valTextInside
      }
    ]
  })
}

const useBarStyle = (percentage: Ref<number>, color: Ref<AnyColor>) => {
  return computed(() => {
    const pv = autoFixPercentage(unref(percentage))
    const cv = unref(color)
    const backgroundColor = getColorBy(cv, pv)
    return { width: `${pv}%`, backgroundColor }
  })
}

const useBarOuterStyle = (strokeWidth: Ref<number>) => {
  return computed(() => {
    const sw = unref(strokeWidth)
    return { height: sw + 'px' }
  })
}

const useContent = (format: Ref<FnFormat>, percentage: Ref<number>) => {
  return computed(() => {
    const fv = unref(format)
    const pv = autoFixPercentage(unref(percentage))
    if (format) {
      return fv(pv) || ''
    }
    return `${pv}%`
  })
}

const useIconClass = (status: Ref<string>, type: Ref<string>) => {
  return computed(() => {
    const st = unref(status)
    const t = unref(type) === 'line' ? 'lineIconClass' : 'arcIconClass'
    const stat = STATUS_SETTING[st]
    return (stat && stat[t]) || ''
  })
}

const useCircleStyle = (width: Ref<number>) => {
  return computed(() => {
    const val = unref(width) + 'px'
    return { width: val, height: val }
  })
}

const useSvgStrokeWidth = (strokeWidth: Ref<number>, width: Ref<number>) => {
  return computed(() => {
    const sw = unref(strokeWidth)
    const w = unref(width)
    return calcRelativeSvgSize(sw, w)
  })
}

const useSvgPathD = (
  svgStrokeWidth: ComputedRef<number>,
  type: Ref<string>
) => {
  return computed(() => {
    const ssw = unref(svgStrokeWidth)
    const tv = unref(type)
    return generateSvgPathD(ssw, tv)
  })
}

const useTrailPathStyle = (
  svgStrokeWidth: ComputedRef<number>,
  type: Ref<string>
) => {
  return computed(() => {
    const ssw = unref(svgStrokeWidth)
    const radius = calcSvgRadius(ssw)
    const perimeter = calcPerimeter(radius)
    const tv = unref(type)
    return genTrailPathStyle(perimeter, tv)
  })
}

const useArcPathStyle = (
  svgStrokeWidth: ComputedRef<number>,
  percentage: Ref<number>,
  type: Ref<string>
) => {
  return computed(() => {
    const ssw = unref(svgStrokeWidth)
    const radius = calcSvgRadius(ssw)
    const perimeter = calcPerimeter(radius)
    const percent = autoFixPercentage(unref(percentage))
    const tv = unref(type)
    return genArcPathStyle(perimeter, percent, tv)
  })
}

const useSvgStrokeColor = (
  status: Ref<string>,
  color: Ref<AnyColor>,
  percentage: Ref<number>
) => {
  return computed(() => {
    const s = unref(status)
    const c = unref(color)
    const p = autoFixPercentage(unref(percentage))
    return getSvgStrokeColor(s, c, p)
  })
}

/**
 * Only change when type is not 'line'
 */
const useTextStyle = (type: Ref<string>, width: Ref<number>) => {
  return computed(() => {
    const t = unref(type)
    const w = unref(width)
    const fontSize = (w * 0.11 + 2).toFixed() + 'px'
    return t === 'line' ? '' : { fontSize }
  })
}
</script>

<style></style>
