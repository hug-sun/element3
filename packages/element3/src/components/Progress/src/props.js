import { isArray, isNumber, isString, isFunction } from '../../../utils/types'
import { isEmpty } from '../../../utils/util'

export const DEFAULT_COLOR = '#409EFF'
export const STATUS_SETTING = {
  success: {
    color: '#67c23a',
    lineIconClass: 'el-icon-circle-check',
    arcIconClass: 'el-icon-check'
  },
  warning: {
    color: '#e6a23c',
    lineIconClass: 'el-icon-warning',
    arcIconClass: 'el-icon-warning'
  },
  exception: {
    color: '#f56c6c',
    lineIconClass: 'el-icon-circle-close',
    arcIconClass: 'el-icon-close'
  }
}
export const STATUSES = Object.keys(STATUS_SETTING)
export const TYPES = ['line', 'circle', 'dashboard']
export const LINECAPS = ['butt', 'round', 'square']

export const statusValid = (val) =>
  isEmpty(val) || (!isEmpty(val) && STATUSES.includes(val))
export const percentageValid = (val) => isNumber(val) && val >= 0 && val <= 100
export const typeValid = (val) => TYPES.includes(val)
export const linecapValid = (val) => LINECAPS.includes(val)

export const FULL_PERCENT = 100
export const DEFAULT_SVG_PX = 126
export const DEFAULT_STROKE_PX = 6
export const SVG_MAX_SIZE = 100
export const SVG_VIEW_BOX = generateViewBox(SVG_MAX_SIZE)
export const DEFAULT_FIXED = 1
export const RATE_CIRCLE = 1
export const DASHBOARD_RATE = 0.75
export const TRANSITION = 'stroke-dasharray 0.6s ease 0s, stroke 0.6s ease 0s'

export const props = {
  type: {
    type: String,
    default: 'line',
    validator: typeValid
  },
  percentage: {
    type: Number,
    default: 0,
    required: true,
    validator: percentageValid
  },
  format: Function,
  status: {
    type: String,
    default: '',
    required: false,
    validator: statusValid
  },
  color: { type: [String, Function, Array], default: '' },
  showText: {
    type: Boolean,
    default: true
  },
  strokeWidth: {
    type: Number,
    default: DEFAULT_STROKE_PX
  },
  textInside: {
    type: Boolean,
    default: false
  },
  width: {
    type: Number,
    default: DEFAULT_SVG_PX
  },
  strokeLinecap: {
    type: String,
    default: 'round',
    validator: linecapValid
  }
}

export function getColorsIndex(colors, percent) {
  const i = colors.findIndex((c) => percent < c.percentage)
  return i < 0 ? colors.length - 1 : i
}

export function sortByPercentage(pre, next) {
  return pre.percentage - next.percentage
}

export function toPercentageColors(colors) {
  const span = FULL_PERCENT / colors.length
  return colors.map((color, i) => {
    if (isString(color)) {
      return { color, percentage: span * (i + 1) }
    }
    return color
  })
}

export function autoFixPercentage(percentage) {
  if (percentage < 0) {
    return 0
  }
  if (percentage > FULL_PERCENT) {
    return FULL_PERCENT
  }
  return percentage
}

export function generateViewBox(size) {
  return `0 0 ${size} ${size}`
}

export function generateSvgPathD(strokeWidth, type) {
  const half = SVG_MAX_SIZE / 2
  const radius = calcSvgRadius(strokeWidth)
  const diameter = radius * 2
  const isDashboard = type === 'dashboard'
  const fromTo = isDashboard ? '' : '-'
  const toFrom = isDashboard ? '-' : ''
  const d = `M ${half} ${half} m 0 ${fromTo}${radius} a ${radius} ${radius} 0 1 1 0 ${toFrom}${diameter} a ${radius} ${radius} 0 1 1 0 ${fromTo}${diameter}`
  return d
}

export function genFnToRelativeSvgSize(width) {
  return (size) => {
    return (size / width) * SVG_MAX_SIZE
  }
}

export function toFixedStr(f) {
  return f.toFixed(DEFAULT_FIXED)
}

export function calcRelativeSvgSize(size, width) {
  return toFixedStr(genFnToRelativeSvgSize(width)(size))
}

export function calcSvgRadius(strokeWidth) {
  return SVG_MAX_SIZE / 2 - strokeWidth / 2
}

export function calcPerimeter(radius) {
  return 2 * Math.PI * radius
}

export function genTrailPathStyle(perimeter, type = 'circle') {
  const rate = getRate(type)
  const offset = toFixedStr(getOffset(perimeter, rate))
  const range = toFixedStr(perimeter * rate)
  const all = toFixedStr(perimeter)
  const strokeDasharray = `${range}px, ${all}px`
  const strokeDashoffset = `${offset}px`
  return { strokeDasharray, strokeDashoffset }
}

export function getRate(type) {
  return type === 'dashboard' ? DASHBOARD_RATE : 1
}

export function genArcPathStyle(perimeter, percentage = 0, type = 'circle') {
  const rate = getRate(type)
  const offset = toFixedStr(getOffset(perimeter, rate))
  const p = toFixedStr(perimeter * (percentage / FULL_PERCENT) * rate)
  const s = toFixedStr(perimeter)
  const strokeDasharray = `${p}px, ${s}px`
  const strokeDashoffset = `${offset}px`
  const transition = TRANSITION
  return { strokeDasharray, strokeDashoffset, transition }
}

export function getSvgStrokeColor(status, color, percentage) {
  if (!isEmpty(color)) {
    return getColorBy(color, percentage)
  }
  if (!isEmpty(status)) {
    return STATUS_SETTING[status].color
  }
  return DEFAULT_COLOR
}

export function getOffset(perimeter, rate) {
  return (-1 * perimeter * (1 - rate)) / 2
}

export function getColorBy(color, percentage) {
  if (isArray(color)) {
    const cs = toPercentageColors(color).sort(sortByPercentage)
    const i = getColorsIndex(cs, percentage)
    return cs[i].color
  }
  if (isFunction(color)) {
    return color(percentage)
  }
  if (isString(color)) {
    return color
  }
}
