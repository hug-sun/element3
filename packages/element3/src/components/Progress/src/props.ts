import { isArray, isNumber, isString, isFunction } from '../../../utils/types'
import { PropType } from 'vue'
import { isEmpty } from '../../../utils/util'
import {
  AnyColor,
  FnColor,
  LevelColor,
  ProgressProps,
  StrokeStyle,
  FnFormat
} from './types'

export const DEFAULT_COLOR = '#409EFF'
export const STATUS_SETTING = {
  success: {
    color: '#67c23a',
    lineIconClass: 'el-icon-circle-check',
    arcIconClass: 'el-icon-check'
  },
  warning: {
    color: '#e6a23c',
    lineIconClass: 'el-icon-warning',
    arcIconClass: 'el-icon-warning'
  },
  exception: {
    color: '#f56c6c',
    lineIconClass: 'el-icon-circle-close',
    arcIconClass: 'el-icon-close'
  }
}
export const STATUSES = Object.keys(STATUS_SETTING)
export const TYPES = ['line', 'circle', 'dashboard']
export const LINECAPS = ['butt', 'round', 'square']

export const statusValid = (val: string): boolean =>
  isEmpty(val) || (!isEmpty(val) && STATUSES.includes(val))
export const percentageValid = (val: number): boolean =>
  isNumber(val) && val >= 0 && val <= 100
export const typeValid = (val: string): boolean => TYPES.includes(val)
export const linecapValid = (val: string): boolean => LINECAPS.includes(val)

export const FULL_PERCENT = 100
export const DEFAULT_SVG_PX = 126
export const DEFAULT_STROKE_PX = 6
export const SVG_MAX_SIZE = 100
export const SVG_VIEW_BOX = generateViewBox(SVG_MAX_SIZE)
export const DEFAULT_FIXED = 1
export const RATE_CIRCLE = 1
export const DASHBOARD_RATE = 0.75
export const TRANSITION = 'stroke-dasharray 0.6s ease 0s, stroke 0.6s ease 0s'

export const props = {
  type: {
    type: String,
    default: 'line',
    validator: typeValid
  },
  percentage: {
    type: Number,
    default: 0,
    required: true,
    validator: percentageValid
  },
  format: {
    type: Function as PropType<FnFormat>
  },
  status: {
    type: String,
    default: '',
    required: false,
    validator: statusValid
  },
  color: {
    type: [String, Function, Array] as PropType<AnyColor>,
    default: ''
  },
  showText: {
    type: Boolean,
    default: true
  },
  strokeWidth: {
    type: Number,
    default: DEFAULT_STROKE_PX
  },
  textInside: {
    type: Boolean,
    default: false
  },
  width: {
    type: Number,
    default: DEFAULT_SVG_PX
  },
  strokeLinecap: {
    type: String,
    default: 'round',
    validator: linecapValid
  }
}

export function getColorsIndex(colors: LevelColor[], percent: number): number {
  const i = colors.findIndex((c) => percent < c.percentage)
  return i < 0 ? colors.length - 1 : i
}

export function sortByPercentage(
  pre: ProgressProps,
  next: ProgressProps
): number {
  return pre.percentage - next.percentage
}

export function toPercentageColors(colors: any[]): any[] {
  const span = FULL_PERCENT / colors.length
  return colors.map((color, i) => {
    if (isString(color)) {
      return { color, percentage: span * (i + 1) }
    }
    return color
  })
}

export function autoFixPercentage(percentage: number): number {
  if (percentage < 0) {
    return 0
  }
  if (percentage > FULL_PERCENT) {
    return FULL_PERCENT
  }
  return percentage
}

export function generateViewBox(size: number): string {
  return `0 0 ${size} ${size}`
}

export function generateSvgPathD(strokeWidth: number, type?: string): string {
  const half = SVG_MAX_SIZE / 2
  const radius = calcSvgRadius(strokeWidth)
  const diameter = radius * 2
  const isDashboard = type === 'dashboard'
  const fromTo = isDashboard ? '' : '-'
  const toFrom = isDashboard ? '-' : ''
  const d = `M ${half} ${half} m 0 ${fromTo}${radius} a ${radius} ${radius} 0 1 1 0 ${toFrom}${diameter} a ${radius} ${radius} 0 1 1 0 ${fromTo}${diameter}`
  return d
}

export function genFnToRelativeSvgSize(
  width: number
): (size: number) => number {
  return (size: number): number => {
    return (size / width) * SVG_MAX_SIZE
  }
}

export function toFixedStr(f: number): string {
  return f.toFixed(DEFAULT_FIXED)
}

export function calcRelativeSvgSize(size: number, width: number): number {
  return Number.parseFloat(toFixedStr(genFnToRelativeSvgSize(width)(size)))
}

export function calcSvgRadius(strokeWidth: number): number {
  return SVG_MAX_SIZE / 2 - strokeWidth / 2
}

export function calcPerimeter(radius: number): number {
  return 2 * Math.PI * radius
}

export function genTrailPathStyle(
  perimeter: number,
  type = 'circle'
): StrokeStyle {
  const rate = getRate(type)
  const offset = toFixedStr(getOffset(perimeter, rate))
  const range = toFixedStr(perimeter * rate)
  const all = toFixedStr(perimeter)
  const strokeDasharray = `${range}px, ${all}px`
  const strokeDashoffset = `${offset}px`
  return { strokeDasharray, strokeDashoffset }
}

export function getRate(type: string): number {
  return type === 'dashboard' ? DASHBOARD_RATE : 1
}

export function genArcPathStyle(
  perimeter: number,
  percentage = 0,
  type = 'circle'
): StrokeStyle {
  const rate = getRate(type)
  const offset = toFixedStr(getOffset(perimeter, rate))
  const p = toFixedStr(perimeter * (percentage / FULL_PERCENT) * rate)
  const s = toFixedStr(perimeter)
  const strokeDasharray = `${p}px, ${s}px`
  const strokeDashoffset = `${offset}px`
  const transition = TRANSITION
  return { strokeDasharray, strokeDashoffset, transition }
}

export function getSvgStrokeColor(
  status?: string,
  color?: AnyColor,
  percentage?: number
): AnyColor {
  if (!isEmpty(color)) {
    return getColorBy(color, percentage)
  }
  if (!isEmpty(status)) {
    return STATUS_SETTING[status].color
  }
  return DEFAULT_COLOR
}

export function getOffset(perimeter: number, rate: number): number {
  return (-1 * perimeter * (1 - rate)) / 2
}

export function getColorBy(color: AnyColor, percentage: number): AnyColor {
  if (isArray(color)) {
    const colors = color as LevelColor[]
    const cs = toPercentageColors(colors).sort(sortByPercentage)
    const i = getColorsIndex(cs, percentage)
    return cs[i].color
  }
  if (isFunction(color)) {
    const fnColor = color as FnColor
    return fnColor(percentage)
  }
  if (isString(color)) {
    return color
  }
}
