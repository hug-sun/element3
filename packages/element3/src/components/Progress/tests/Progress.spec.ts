import { DEFAULT_COLOR, sortByPercentage } from '../src/props'
import {
  expectHaveClass,
  expectHaveTextContent,
  expectNotHaveAttribute,
  expectNotHaveClass,
  expectNotToBeExist,
  expectToBeExist
} from '../../../../tests/helper'

import {
  initProgress,
  assertSetPercentage,
  assertSetBgColor,
  assertNotContainStyle,
  assertContainStyle,
  assertArcStyleOk,
  findSvgTrailPath,
  findSvgArcPath,
  assertSvgStrokeOk,
  assertIconClassOk
} from './test-helper'

const colorPercents = [
  { color: '#1989fa', percentage: 80 },
  { color: '#6f7ad3', percentage: 100 },
  { color: '#5cb87a', percentage: 60 },
  { color: '#f56c6c', percentage: 20 },
  { color: '#e6a23c', percentage: 40 }
]
colorPercents.sort(sortByPercentage)

describe('Progress.vue', () => {
  it('snapshot', () => {
    const wrapper = initProgress()
    expect(wrapper.element).toMatchSnapshot()
  })
  it('should create default Progress component and HTML structure', () => {
    const percentage = 55
    const wrapper = initProgress({ percentage })
    expectHaveClass(wrapper, 'el-progress')
    expectHaveClass(wrapper, 'el-progress--line')
    expectNotHaveClass(wrapper, 'is-undefined')

    expectToBeExist(wrapper.find('.el-progress > .el-progress-bar'))
    expectToBeExist(wrapper.find('.el-progress > .el-progress__text'))
    expectToBeExist(wrapper.find('.el-progress-bar > .el-progress-bar__outer'))

    expectToBeExist(
      wrapper.find('.el-progress-bar__outer .el-progress-bar__inner')
    )

    assertSetPercentage(wrapper, percentage)
  })

  describe('line type progress:', () => {
    it('percentage', async () => {
      const percentage = 58
      const wrapper = initProgress({ percentage })
      assertSetPercentage(wrapper, 58)
      await wrapper.setProps({ percentage: 28 })
      assertSetPercentage(wrapper, 28)
    })

    it('percentage validator', async () => {
      const percentage = 158
      const wrapper = initProgress({ percentage })
      assertSetPercentage(wrapper, 100)
      await wrapper.setProps({ percentage: -28 })
      assertSetPercentage(wrapper, 0)
    })

    it('format', () => {
      const format = (p) => (p === 100 ? '满' : `${p}%`)
      const percentage = 100
      const props = { percentage, format }
      const wrapper = initProgress(props)
      expectHaveTextContent(wrapper.get('.el-progress__text'), '满')
    })

    it('color string', async () => {
      const color1 = '#409eff'
      const percentage = 30
      const props = { percentage, color: color1 }
      const wrapper = initProgress(props)
      assertSetBgColor(wrapper, color1)
      const color2 = '#336699'
      await wrapper.setProps({ color: color2 })
      assertSetBgColor(wrapper, color2)
      const color3 = ''
      await wrapper.setProps({ color: color3 })
      assertNotContainStyle(
        wrapper,
        '.el-progress-bar__inner',
        'background-color'
      )
    })

    it('color function', async () => {
      const color = (percentage: number) => {
        if (percentage < 30) {
          return '#33ff99'
        } else if (percentage < 70) {
          return '#ff9900'
        } else {
          return '#993300'
        }
      }
      const p1 = 20
      const props = { percentage: p1, color }
      const wrapper = initProgress(props)

      assertSetBgColor(wrapper, color(p1))

      const p2 = 50
      await wrapper.setProps({ percentage: p2 })

      assertSetBgColor(wrapper, color(p2))

      const p3 = 80
      await wrapper.setProps({ percentage: p3 })

      assertSetBgColor(wrapper, color(p3))
    })

    it('color string array', async () => {
      const colors = ['#336699', '#339966', '#996633']
      const percentage = 15
      const props = { percentage, color: colors }
      const wrapper = initProgress(props)
      assertSetBgColor(wrapper, colors[0])

      await wrapper.setProps({ percentage: 0 })
      assertSetBgColor(wrapper, colors[0])
      await wrapper.setProps({ percentage: 10 })
      assertSetBgColor(wrapper, colors[0])
      await wrapper.setProps({ percentage: 35 })
      assertSetBgColor(wrapper, colors[1])
      await wrapper.setProps({ percentage: 65 })
      assertSetBgColor(wrapper, colors[1])
      await wrapper.setProps({ percentage: 67 })
      assertSetBgColor(wrapper, colors[2])
      await wrapper.setProps({ percentage: 100 })
      assertSetBgColor(wrapper, colors[2])
    })

    it('color object array', () => {
      expect(colorPercents[0].percentage).toBe(20)
      expect(colorPercents[3].percentage).toBe(80)
      expect(colorPercents[4].percentage).toBe(100)

      const wrapper = initProgress({ color: colorPercents })
      expect(wrapper).toBeDefined()
      assertSetBgColor(wrapper, colorPercents[4].color)
    })

    it('status add "is-xxx" class', async () => {
      const wrapper = initProgress({ status: 'success' })
      expectHaveClass(wrapper, 'is-success')

      expectToBeExist(
        wrapper.find('.el-progress__text > i.el-icon-circle-check')
      )

      await wrapper.setProps({ status: 'exception' })
      expectHaveClass(wrapper, 'is-exception')

      expectToBeExist(
        wrapper.find('.el-progress__text > i.el-icon-circle-close')
      )
      await wrapper.setProps({ status: 'warning' })
      expectHaveClass(wrapper, 'is-warning')
      expectToBeExist(wrapper.find('.el-progress__text > i.el-icon-warning'))

      await wrapper.setProps({ status: 'error' })
      expectNotHaveClass(wrapper, 'is-error')
    })

    it('show-text', async () => {
      const wrapper = initProgress()
      expectToBeExist(wrapper.find('.el-progress__text'))
      expectHaveTextContent(wrapper.get('.el-progress__text'), '85%')

      await wrapper.setProps({ showText: false })
      expectNotToBeExist(wrapper.find('.el-progress__text'))
      expectNotToBeExist(wrapper.find('.el-progress-bar__innerText'))
      expectHaveClass(wrapper, 'el-progress--without-text')
    })

    it('strokeWidth', async () => {
      const wrapper = initProgress({ strokeWidth: 30 })
      assertContainStyle(
        wrapper,
        '.el-progress-bar > .el-progress-bar__outer',
        'height: 30px'
      )
      await wrapper.setProps({ strokeWidth: 20 })
      assertContainStyle(
        wrapper,
        '.el-progress-bar > .el-progress-bar__outer',
        'height: 20px'
      )
    })

    it('textInside', async () => {
      const wrapper = initProgress()
      expectNotHaveClass(wrapper, 'el-progress--text-inside')
      expectToBeExist(
        wrapper.find('.el-progress-bar__outer > .el-progress-bar__inner')
      )
      expectNotToBeExist(wrapper.find('.el-progress-bar__innerText'))

      await wrapper.setProps({ textInside: true })
      expectHaveClass(wrapper, 'el-progress--text-inside')
      expectToBeExist(wrapper.find('.el-progress-bar__innerText'))
      expectNotToBeExist(wrapper.find('.el-progress__text'))
    })
  })

  describe('circle type progress', () => {
    it('type prop', () => {
      const wrapper = initProgress({ type: 'circle' })
      expectNotHaveClass(wrapper, 'el-progress--line')
      expectHaveClass(wrapper, 'el-progress--circle')
      expectNotToBeExist(wrapper.find('.el-progress-bar'))
      expectToBeExist(wrapper.find('.el-progress-circle'))
    })

    it('circle html and svg etc', () => {
      const wrapper = initProgress({ type: 'circle' })
      expectToBeExist(wrapper.find('.el-progress-circle > svg'))
      const circle = wrapper.find('.el-progress-circle')
      expect(circle.attributes().style).toBe('width: 126px; height: 126px;')
      const svg = wrapper.find('.el-progress-circle > svg')
      expect(svg.attributes().viewBox).toBe('0 0 100 100')
      const svgTrailPath = findSvgTrailPath(wrapper)
      expectHaveClass(svgTrailPath, 'el-progress-circle__track')

      const svgArcPath = findSvgArcPath(wrapper)
      expectHaveClass(svgArcPath, 'el-progress-circle__path')

      const d = `M 50 50 m 0 -47.6 a 47.6 47.6 0 1 1 0 95.2 a 47.6 47.6 0 1 1 0 -95.2`
      const trailAttrs = svgTrailPath.attributes()
      expect(trailAttrs.d).toContain(d)
      expect(trailAttrs['stroke-width']).toBe('4.8')
      expect(trailAttrs.style).toBe(
        'stroke-dasharray: 299.1px, 299.1px; stroke-dashoffset: 0.0px;'
      )
      assertArcStyleOk(wrapper, 85)
    })

    it('percentage', async () => {
      const wrapper = initProgress({ type: 'circle', percentage: 25 })
      await assertArcStyleOk(wrapper)
      await assertArcStyleOk(wrapper, 50)
    })

    it('strokeLinecap', async () => {
      const wrapper = initProgress({ type: 'circle' })
      const svgArcPath = findSvgArcPath(wrapper)
      expect(svgArcPath.attributes()['stroke-linecap']).toBe('round')

      await wrapper.setProps({ strokeLinecap: 'butt' })
      expect(svgArcPath.attributes()['stroke-linecap']).toBe('butt')
      await wrapper.setProps({ strokeLinecap: 'square' })
      expect(svgArcPath.attributes()['stroke-linecap']).toBe('square')
    })

    it('default color stroke', async () => {
      const wrapper = initProgress({ type: 'circle' })
      const svgArcPath = findSvgArcPath(wrapper)
      expect(svgArcPath.attributes()['stroke']).toBe(DEFAULT_COLOR)
    })

    it('status change stroke color', async () => {
      const wrapper = initProgress({ type: 'circle' })
      let status = 'success'
      await wrapper.setProps({ status })
      assertSvgStrokeOk(wrapper, status)

      status = 'warning'
      await wrapper.setProps({ status })
      assertSvgStrokeOk(wrapper, status)

      status = 'exception'
      await wrapper.setProps({ status })
      assertSvgStrokeOk(wrapper, status)
    })

    it('icon class', async () => {
      const wrapper = initProgress({ type: 'circle' })
      let status = 'success'
      await wrapper.setProps({ status })
      assertIconClassOk(wrapper, status)

      status = 'warning'
      await wrapper.setProps({ status })
      assertIconClassOk(wrapper, status)

      status = 'exception'
      await wrapper.setProps({ status })
      assertIconClassOk(wrapper, status)
    })

    it('font size', async () => {
      const wrapper = initProgress()
      expectNotHaveAttribute(wrapper.get('.el-progress__text'), 'style')
      await wrapper.setProps({ strokeWidth: 50 })
      expectNotHaveAttribute(wrapper.get('.el-progress__text'), 'style')
      await wrapper.setProps({ type: 'circle' })
      assertContainStyle(wrapper, '.el-progress__text', 'font-size: 16px;')
      await wrapper.setProps({ width: 200 })
      assertContainStyle(wrapper, '.el-progress__text', 'font-size: 24px;')
    })
  })

  describe('dashboard type progress', () => {
    it('dashboard progress', async () => {
      const wrapper = initProgress({ type: 'dashboard' })
      const svgArcPath = findSvgArcPath(wrapper)
      expect(svgArcPath.attributes()['stroke']).toBe(DEFAULT_COLOR)
    })

    it('dashboard progress with color', async () => {
      const wrapper = initProgress({ type: 'dashboard', color: colorPercents })
      const svgTrailPath = findSvgTrailPath(wrapper)
      const trailAttrs = svgTrailPath.attributes()
      expect(trailAttrs.style).toBe(
        'stroke-dasharray: 224.3px, 299.1px; stroke-dashoffset: -37.4px;'
      )

      const d = `M 50 50 m 0 47.6 a 47.6 47.6 0 1 1 0 -95.2 a 47.6 47.6 0 1 1 0 95.2`
      expect(trailAttrs.d).toContain(d)

      const svgArcPath = findSvgArcPath(wrapper)
      expect(svgArcPath.attributes()['stroke']).toBe(colorPercents[4].color)
      const arcAttrs = svgArcPath.attributes()
      expect(arcAttrs.style).toBe(
        'stroke-dasharray: 190.7px, 299.1px; stroke-dashoffset: -37.4px; transition: stroke-dasharray 0.6s ease 0s, stroke 0.6s ease 0s;'
      )
    })
  })
})
