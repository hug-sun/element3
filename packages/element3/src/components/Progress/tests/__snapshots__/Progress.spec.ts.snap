// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress.vue snapshot 1`] = `
<div
  class="el-progress el-progress--line"
>
  <div
    class="el-progress-bar"
  >
    <div
      class="el-progress-bar__outer"
      style="height: 6px;"
    >
      <div
        class="el-progress-bar__inner"
        style="width: 85%;"
      >
        <!--v-if-->
      </div>
    </div>
  </div>
  <div
    class="el-progress__text"
  >
    
    85%
    
  </div>
</div>
`;
